### PR TITLE
Ensure that FormattableString is preferred by Introduce Explaining Variable and Extract Method

### PR DESCRIPTION
**Pull request by DustinCampbell**
Fixes Issue #3147

When extracting an interpolated string that is target-typed to FormattableString, the resulting declaration should use FormattableString as the type rather than string. Otherwise, the result is uncompilable code.

Consider the following code:

```C#
void M()
{
    var s = System.FormattableString.Invariant($"");
}
```

If ```$""``` is selected, Introduce Explaining Variable should produce this:

```C#
void M()
{
    System.FormattableString v = $"";
    var s = System.FormattableString.Invariant(v);
}
```

...rather than this:

```C#
void M()
{
    var v = $"";
    var s = System.FormattableString.Invariant(v);
}
```

Added unit tests for both Introduce Explaining Variable and Extract Method in both C# and VB.

Note that this really isn't the *correct* fix for Introduce Explaining Variable. There should be a simplification step that determines whether a local declaration's type is changed to var (respecting the user's options), but that's not how it was implemented. However, doing that is a larger change that is probably not appropriate for the stabilization branch. So, there are some edge cases that will slip through with this change, but they are not as functionally broken as the case this change fixes.

### CODE REVIEW

**Comment by DustinCampbell** [comment]
I'll be looking for a little love from @Pilchie, @balajikris, @dpoeschl, @rchande, @brettfo, @jasonmalinowski.

File: src/Features/Core/IntroduceVariable/AbstractIntroduceVariableService.cs [hunk 1]
```diff
@@ -310,6 +310,12 @@ where NodeMatchesExpression(originalSemanticModel, currentSemanticModel, syntaxF
             var semanticModel = document.SemanticModel;
             var typeInfo = semanticModel.GetTypeInfo(expression, cancellationToken);
 
+            if (typeInfo.Type?.SpecialType == SpecialType.System_String &&
+                typeInfo.ConvertedType.IsFormattableString())
```
Comment (brettfo) [inline src/Features/Core/IntroduceVariable/AbstractIntroduceVariableService.cs]: Can `typeinfo.ConvertedType` ever be null?  Other places you do similar things you've used Elvis.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs [hunk 2]
```diff
@@ -368,6 +368,14 @@ public static bool IsAttribute(this ITypeSymbol symbol)
             return false;
         }
 
+        public static bool IsFormattableString(this ITypeSymbol symbol)
+        {
+            return symbol?.MetadataName == "FormattableString"
+                && symbol.ContainingNamespace?.Name == "System"
+                && symbol.ContainingNamespace.ContainingNamespace?.IsGlobalNamespace == true;
+
```
Comment (brettfo) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: Nit: newline.

Comment (DustinCampbell) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: Good catch. Missed that -- thanks!

File: src/Features/Core/IntroduceVariable/AbstractIntroduceVariableService.cs (hunk 1 above)
Comment (DustinCampbell) [inline src/Features/Core/IntroduceVariable/AbstractIntroduceVariableService.cs]: Great catch! Will fix.

File: src/EditorFeatures/CSharpTest/Utilities/CodeSnippets.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+﻿namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests
+{
+    internal static class CodeSnippets
+    {
+        public const string FormattableStringType = @"
+namespace System
+{
+    public abstract class FormattableString : IFormattable
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/Utilities/CodeSnippets.cs]: Should we just move our unit tests forward to test against .NET 4.6?

**Comment by jasonmalinowski** [comment]
:heart: :+1:

File: src/EditorFeatures/CSharpTest/Utilities/CodeSnippets.cs (hunk 3 above)
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/Utilities/CodeSnippets.cs]: Today, we load specific assemblies out of our test resources. Updating those to 4.6 feels out of scope for this change. (Note that the compiler tests do the same trick we're doing here.)

**Comment by dpoeschl** [comment]
:cupid: :+1:

File: src/EditorFeatures/CSharpTest/Utilities/CodeSnippets.cs (hunk 3 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/Utilities/CodeSnippets.cs]: Ah, forgot we did that. Might be a good bug for some post-1.0 work.

**Comment by Pilchie** [comment]
:+1:

**Comment by MattGertz** [comment]
Approved 5/29 by ML Shiproom.
